File: src/EnderClient.Core/Game/Injector.cs
using System.Text;
using EnderClient.Core.Exceptions;
using EnderClient.Core.Native;

namespace EnderClient.Core.Game;

/// <summary>
/// Injects a DLL into a running process via the classic LoadLibraryW + CreateRemoteThread technique.
/// </summary>
public static class Injector
{
    public static void Inject(int processId, string dllPath)
    {
        if (!File.Exists(dllPath))
            throw new InjectionFailedException($"DLL not found: {dllPath}");

        var kernel32 = NativeMethods.GetModuleHandle("kernel32.dll");
        var loadLibraryAddr = NativeMethods.GetProcAddress(kernel32, "LoadLibraryW");
        if (loadLibraryAddr == IntPtr.Zero)
            throw new InjectionFailedException("Could not resolve LoadLibraryW.");

        var hProcess = NativeMethods.OpenProcess(NativeMethods.PROCESS_ALL_ACCESS, false, (uint)processId);
        if (hProcess == IntPtr.Zero)
            throw new InjectionFailedException("Failed to open the Minecraft process. Try running Ender Client as Administrator.");

        var pathBytes = Encoding.Unicode.GetBytes(dllPath + '\0');
        var remoteBuffer = IntPtr.Zero;
        var hThread = IntPtr.Zero;

        try
        {
            remoteBuffer = NativeMethods.VirtualAllocEx(hProcess, IntPtr.Zero, (nuint)pathBytes.Length,
                NativeMethods.MEM_COMMIT | NativeMethods.MEM_RESERVE, NativeMethods.PAGE_READWRITE);

            if (remoteBuffer == IntPtr.Zero)
                throw new InjectionFailedException("Failed to allocate memory in the target process.");

            if (!NativeMethods.WriteProcessMemory(hProcess, remoteBuffer, pathBytes, (nuint)pathBytes.Length, out _))
                throw new InjectionFailedException("Failed to write the DLL path into the target process.");

            hThread = NativeMethods.CreateRemoteThread(hProcess, IntPtr.Zero, 0, loadLibraryAddr, remoteBuffer, 0, out _);
            if (hThread == IntPtr.Zero)
                throw new InjectionFailedException("Failed to create the remote thread for injection.");

            NativeMethods.WaitForSingleObject(hThread, NativeMethods.INFINITE);
        }
        finally
        {
            if (hThread != IntPtr.Zero) NativeMethods.CloseHandle(hThread);
            if (remoteBuffer != IntPtr.Zero) NativeMethods.VirtualFreeEx(hProcess, remoteBuffer, 0, NativeMethods.MEM_RELEASE);
            NativeMethods.CloseHandle(hProcess);
        }
    }
}

File: src-tauri/src/injector.rs
use std::ffi::c_void;
use windows::core::{PCSTR, PCWSTR};
use windows::Win32::Foundation::CloseHandle;
use windows::Win32::System::Diagnostics::Debug::WriteProcessMemory;
use windows::Win32::System::LibraryLoader::{GetModuleHandleW, GetProcAddress};
use windows::Win32::System::Memory::{
    VirtualAllocEx, VirtualFreeEx, MEM_COMMIT, MEM_RELEASE, MEM_RESERVE, PAGE_READWRITE,
};
use windows::Win32::System::Threading::{
    CreateRemoteThread, OpenProcess, WaitForSingleObject, INFINITE, LPTHREAD_START_ROUTINE,
    PROCESS_ALL_ACCESS,
};

/// Injects a DLL into a running process via the classic LoadLibraryW + CreateRemoteThread
/// technique — the same approach used by the original C# Injector. LoadLibraryW's own
/// signature (`extern "system" fn(*mut c_void) -> u32` once resolved) already matches
/// LPTHREAD_START_ROUTINE, so no double-transmute through an intermediate integer type
/// is needed — a single transmute of the resolved FARPROC is enough.
pub fn inject(process_id: u32, dll_path: &str) -> Result<(), String> {
    if !std::path::Path::new(dll_path).exists() {
        return Err(format!("DLL not found: {dll_path}"));
    }

    unsafe {
        let kernel32_name: Vec<u16> = "kernel32.dll\0".encode_utf16().collect();
        let kernel32 = GetModuleHandleW(PCWSTR(kernel32_name.as_ptr()))
            .map_err(|e| format!("Failed to get kernel32.dll handle: {e}"))?;

        let load_library_addr = GetProcAddress(kernel32, PCSTR(b"LoadLibraryW\0".as_ptr()))
            .ok_or_else(|| "Could not resolve LoadLibraryW.".to_string())?;

        let start_routine: LPTHREAD_START_ROUTINE = std::mem::transmute(load_library_addr);

        let process = OpenProcess(PROCESS_ALL_ACCESS, false, process_id).map_err(|_| {
            "Failed to open the Minecraft process. Try running Ender Client as Administrator."
                .to_string()
        })?;

        let path_bytes: Vec<u16> = dll_path.encode_utf16().chain(std::iter::once(0)).collect();
        let size = path_bytes.len() * std::mem::size_of::<u16>();

        let remote_buffer = VirtualAllocEx(process, None, size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);

        if remote_buffer.is_null() {
            let _ = CloseHandle(process);
            return Err("Failed to allocate memory in the target process.".to_string());
        }

        let write_result = WriteProcessMemory(
            process,
            remote_buffer,
            path_bytes.as_ptr() as *const c_void,
            size,
            None,
        );

        if write_result.is_err() {
            let _ = VirtualFreeEx(process, remote_buffer, 0, MEM_RELEASE);
            let _ = CloseHandle(process);
            return Err("Failed to write the DLL path into the target process.".to_string());
        }

        let thread = CreateRemoteThread(
            process,
            None,
            0,
            start_routine,
            Some(remote_buffer as *const c_void),
            0,
            None,
        );

        let thread = match thread {
            Ok(h) => h,
            Err(_) => {
                let _ = VirtualFreeEx(process, remote_buffer, 0, MEM_RELEASE);
                let _ = CloseHandle(process);
                return Err("Failed to create the remote thread for injection.".to_string());
            }
        };

        WaitForSingleObject(thread, INFINITE);

        let _ = CloseHandle(thread);
        let _ = VirtualFreeEx(process, remote_buffer, 0, MEM_RELEASE);
        let _ = CloseHandle(process);
    }

    Ok(())
}
